File: frontend/lib/aiRecommendations.ts
/**
 * Compact "So what?" recommendations per sector — derived from live CORS + space weather.
 * Not a separate ML model; rule-based fusion of the same inputs as Navigation News briefs.
 */
import { effectiveNavigationTone } from "./gnssAudienceNews";
import type { ForecastStatus, GnssForecastCity } from "./gnssWeatherIntelligence";
import type { GicStatusResponse, SpaceWeatherCurrent } from "./types";
import {
  formatPowerIndicesDetail,
  formatTelecomIndicesDetail,
  noaaGScaleFromKp,
  formatKpDisplay,
  formatVtecDisplay,
} from "./spaceWeatherMetrics";

export type AiRecommendationAudience =
  | "surveyors"
  | "farmers"
  | "pilots"
  | "power"
  | "telecom";

export interface AiAudienceRecommendation {
  id: AiRecommendationAudience;
  label: string;
  icon: string;
  headline: string;
  detail?: string;
  tone: ForecastStatus;
}

export interface AiRecommendationResult {
  recommendations: AiAudienceRecommendation[];
  tone: ForecastStatus;
  computedAt: string | null;
}

function field(city: GnssForecastCity | undefined, label: string): string | undefined {
  return city?.fields.find((f) => f.label === label)?.value;
}

function surveyorDetail(city: GnssForecastCity | undefined, sw: SpaceWeatherCurrent | null): string | undefined {
  const g = noaaGScaleFromKp(sw?.kp);
  const why = `Why this status? ${g.display} globally · Zimbabwe VTEC ${formatVtecDisplay(sw?.mean_vtec)} · ΔTEC/ROTI pending · provisional GNSS risk ${sw?.gnss_risk ?? "N/A"}`;
  if (!city) return why;
  const accuracy = field(city, "Expected Accuracy");
  const rtk = field(city, "RTK Reliability");
  const window = field(city, "Best Survey Window") ?? "07:00 – 14:00";
  if (!accuracy) {
    return `Survey window ${window} · ${why}`;
  }
  if (city.status === "warning") {
    return `Expected accuracy ${accuracy} · ${why}`;
  }
  const parts = [`Expected accuracy ${accuracy}`];
  if (rtk) parts.push(`RTK ${rtk}`);
  parts.push(`Window ${window}`);
  parts.push(why);
  return parts.join(" · ");
}

function scintillationPilotNote(sw: SpaceWeatherCurrent | null, tone: ForecastStatus): string {
  const s4 = sw?.s4;
  if (tone === "warning" || (s4 != null && s4 >= 0.5)) {
    return "Significant scintillation possible 16:00–21:00 local — monitor RAIM and HF.";
  }
  if (tone === "moderate" || (s4 != null && s4 >= 0.15)) {
    return "Minor scintillation possible after 18:00 local.";
  }
  return "En-route GNSS within normal limits today.";
}

function powerGicNote(sw: SpaceWeatherCurrent | null, gic: GicStatusResponse | null): string {
  const kp = sw?.kp;
  const dst = sw?.dst;
  const g = noaaGScaleFromKp(kp);
  const levels = (gic?.stations ?? [])
    .map((s) => (s.latest_level ?? "").toLowerCase())
    .filter(Boolean);

  if (levels.some((l) => l === "severe" || l === "high" || l === "large")) {
    return "Power utilities — GIC warning: check transformer neutrals and long HV lines.";
  }
  if (
    (kp != null && kp >= 7) ||
    (dst != null && dst <= -100) ||
    levels.some((l) => l === "elevated")
  ) {
    return `Power utilities — elevated GIC watch (${g.display}). Monitor long transmission corridors.`;
  }
  if (g.isStorm || (dst != null && dst <= -50)) {
    return `Power utilities — ${g.display} globally; keep routine GIC watch (geomagnetic ≠ automatic grid fault).`;
  }
  return `Power utilities — ${g.display}; no GIC warning from available monitors.`;
}

function telecomNote(sw: SpaceWeatherCurrent | null, tone: ForecastStatus): string {
  const s4 = sw?.s4;
  const kp = sw?.kp;
  const g = noaaGScaleFromKp(kp);
  if (tone === "warning" || (s4 != null && s4 >= 0.5) || (kp != null && kp >= 6)) {
    return `Telecommunications — timing holds may degrade (${g.display}). Verify GNSS-disciplined clocks and PTP.`;
  }
  if (tone === "moderate" || (s4 != null && s4 >= 0.25)) {
    return `Telecommunications — minor timing jitter possible (${g.display}). GNSS risk remains provisional.`;
  }
  return `Telecommunications — timing stable (${g.display}).`;
}

const SURVEYOR_HEADLINE: Record<ForecastStatus, string> = {
  excellent: "Surveying — GNSS NORMAL. RTK conditions favourable.",
  moderate: "Surveying — GNSS ADVISORY. Allow extra RTK occupation time.",
  warning: "Surveying — GNSS CAUTION. Delay centimetre-critical work.",
};

const FARMER_HEADLINE: Record<ForecastStatus, string> = {
  excellent: "Precision agriculture — GNSS NORMAL. Autosteer favourable.",
  moderate: "Precision agriculture — GNSS ADVISORY. Prefer morning field GPS.",
  warning: "Precision agriculture — GNSS CAUTION. Verify boundaries before legal decisions.",
};

export function buildAiRecommendations(
  forecasts: GnssForecastCity[],
  sw: SpaceWeatherCurrent | null,
  gic: GicStatusResponse | null = null,
  computedAt: string | null = null,
): AiRecommendationResult {
  const tone = effectiveNavigationTone(forecasts, sw);
  const harare = forecasts.find((f) => f.city === "HARARE");
  const mutare = forecasts.find((f) => f.city === "MUTARE");
  const surveyCity = mutare ?? harare;
  const surveyStatus = surveyCity?.status ?? tone;
  const farmerStatus = harare?.status ?? tone;

  const surveyDetail = surveyorDetail(surveyCity, sw);
  const farmerWindow = field(harare, "Best Survey Window");
  const g = noaaGScaleFromKp(sw?.kp);
  const farmerWhy = `Why this status? ${g.display} · VTEC ${formatVtecDisplay(sw?.mean_vtec)} · provisional GNSS risk ${sw?.gnss_risk ?? "N/A"}`;

  const recommendations: AiAudienceRecommendation[] = [
    {
      id: "surveyors",
      label: "Surveying",
      icon: "📐",
      headline: SURVEYOR_HEADLINE[surveyStatus],
      detail: surveyDetail,
      tone: surveyStatus,
    },
    {
      id: "farmers",
      label: "Precision Agriculture",
      icon: "🌾",
      headline: FARMER_HEADLINE[farmerStatus],
      detail:
        farmerStatus === "excellent"
          ? farmerWhy
          : farmerWindow
            ? `Preferred window ${farmerWindow} · ${farmerWhy}`
            : farmerWhy,
      tone: farmerStatus,
    },
    {
      id: "pilots",
      label: "Aviation & Drones",
      icon: "✈️",
      headline:
        tone === "excellent"
          ? "Aviation / drones — GNSS NORMAL. Routine navigation expected."
          : tone === "moderate"
            ? "Aviation / drones — GNSS ADVISORY. Monitor GPS approaches through the afternoon."
            : "Aviation / drones — GNSS CAUTION. Verify navaid backups.",
      detail: `${scintillationPilotNote(sw, tone)} Why this status? ${g.display} (Kp ${formatKpDisplay(sw?.kp)}) · provisional GNSS risk ${sw?.gnss_risk ?? "N/A"} — not automatic from Kp alone.`,
      tone,
    },
    {
      id: "power",
      label: "Power Utilities",
      icon: "⚡",
      headline: powerGicNote(sw, gic),
      detail: formatPowerIndicesDetail(sw),
      tone,
    },
    {
      id: "telecom",
      label: "Telecommunications",
      icon: "📡",
      headline: telecomNote(sw, tone),
      detail: formatTelecomIndicesDetail(sw),
      tone,
    },
  ];

  return {
    recommendations,
    tone,
    computedAt,
  };
}

File: zgiis/navigation/ai_recommendations.py
"""Compact sector recommendations — Python mirror of frontend/lib/aiRecommendations.ts."""
from __future__ import annotations

from typing import Any, Literal

from zgiis.navigation.gnss_forecast import GnssForecastCity
from zgiis.navigation.audience_news import _effective_navigation_tone

AiRecommendationAudience = Literal["surveyors", "farmers", "pilots", "power", "telecom"]

_SURVEYOR_HEADLINE: dict[ForecastStatus, str] = {
    "excellent": "Proceed.",
    "moderate": "Allow extra RTK occupation time.",
    "warning": "Delay centimetre-critical surveys.",
}

_FARMER_HEADLINE: dict[ForecastStatus, str] = {
    "excellent": "Good day for precision planting.",
    "moderate": "Plan GPS-heavy field work before late morning.",
    "warning": "Verify boundaries before legal or financial commitments.",
}


def _field(city: GnssForecastCity | None, label: str) -> str | None:
    if city is None:
        return None
    for f in city.fields:
        if f.get("label") == label:
            return f.get("value")
    return None


def _surveyor_detail(city: GnssForecastCity | None) -> str | None:
    if city is None:
        return None
    accuracy = _field(city, "Expected Accuracy")
    rtk = _field(city, "RTK Reliability")
    window = _field(city, "Best Survey Window") or "07:00 – 14:00"
    if not accuracy:
        return f"Survey window {window}" if window else None
    if city.status == "warning":
        return f"Expected accuracy {accuracy}"
    parts = [f"Expected accuracy {accuracy}"]
    if rtk:
        parts.append(f"RTK {rtk}")
    parts.append(f"Window {window}")
    return " · ".join(parts)


def _scintillation_pilot_note(sw: dict[str, Any] | None, tone: ForecastStatus) -> str:
    s4 = sw.get("s4") if sw else None
    if tone == "warning" or (s4 is not None and s4 >= 0.5):
        return "Significant scintillation possible 16:00–21:00 local — monitor RAIM and HF."
    if tone == "moderate" or (s4 is not None and s4 >= 0.15):
        return "Minor scintillation possible after 18:00 local."
    return "En-route GNSS within normal limits today."


def _power_gic_note(sw: dict[str, Any] | None, gic: dict[str, Any] | None) -> str:
    kp = sw.get("kp") if sw else None
    dst = sw.get("dst") if sw else None
    levels = [
        str(s.get("latest_level") or "").lower()
        for s in (gic or {}).get("stations") or []
        if isinstance(s, dict)
    ]

    if any(l in ("severe", "high", "large") for l in levels):
        return "GIC warning — elevated transformer-neutral currents detected."
    if (
        (kp is not None and kp >= 7)
        or (dst is not None and dst <= -100)
        or any(l == "elevated" for l in levels)
    ):
        return "Elevated GIC risk — increase monitoring on long transmission lines."
    if (kp is not None and kp >= 5) or (dst is not None and dst <= -50):
        return "Routine GIC monitoring — minor geomagnetic disturbance under way."
    return "No GIC warning."


def _telecom_note(sw: dict[str, Any] | None, tone: ForecastStatus) -> str:
    s4 = sw.get("s4") if sw else None
    kp = sw.get("kp") if sw else None
    if tone == "warning" or (s4 is not None and s4 >= 0.5) or (kp is not None and kp >= 6):
        return "Timing holds may degrade — verify GNSS-disciplined clocks and PTP."
    if tone == "moderate" or (s4 is not None and s4 >= 0.25):
        return "Minor timing jitter possible this afternoon on GNSS-linked links."
    return "Timing stable."


def build_ai_recommendations(
    forecasts: list[GnssForecastCity],
    sw: dict[str, Any] | None,
    gic: dict[str, Any] | None = None,
    computed_at: str | None = None,
) -> dict[str, Any]:
    tone = _effective_navigation_tone(forecasts, sw)
    by_city = {f.city: f for f in forecasts}
    harare = by_city.get("HARARE")
    mutare = by_city.get("MUTARE")
    survey_city = mutare or harare
    survey_status = survey_city.status if survey_city else tone
    farmer_status = harare.status if harare else tone

    survey_detail = _surveyor_detail(survey_city)
    farmer_window = _field(harare, "Best Survey Window")

    pilot_headline = {
        "excellent": "Routine GNSS navigation expected.",
        "moderate": "Monitor GPS approaches through the afternoon.",
        "warning": "Storm procedures — verify navaid backups.",
    }[tone]

    recs = [
        {
            "id": "surveyors",
            "label": "Surveyors",
            "icon": "📐",
            "headline": _SURVEYOR_HEADLINE[survey_status],
            "detail": survey_detail,
            "tone": survey_status,
        },
        {
            "id": "farmers",
            "label": "Farmers",
            "icon": "🌾",
            "headline": _FARMER_HEADLINE[farmer_status],
            "detail": None if farmer_status == "excellent" else (f"Preferred window {farmer_window}" if farmer_window else None),
            "tone": farmer_status,
        },
        {
            "id": "pilots",
            "label": "Pilots",
            "icon": "✈️",
            "headline": pilot_headline,
            "detail": _scintillation_pilot_note(sw, tone),
            "tone": tone,
        },
        {
            "id": "power",
            "label": "Power Utilities",
            "icon": "⚡",
            "headline": _power_gic_note(sw, gic),
            "detail": (
                f"Kp {sw['kp']} · Dst {'+' if sw['dst'] >= 0 else ''}{sw['dst']} nT"
                if sw and sw.get("kp") is not None and sw.get("dst") is not None
                else None
            ),
            "tone": tone,
        },
        {
            "id": "telecom",
            "label": "Telecommunications",
            "icon": "📡",
            "headline": _telecom_note(sw, tone),
            "detail": (
                f"S4 {sw['s4']:.2f} · ionospheric amplitude scintillation index"
                if sw and sw.get("s4") is not None
                else None
            ),
            "tone": tone,
        },
    ]

    return {
        "recommendations": recs,
        "tone": tone,
        "computed_at": computed_at,
    }
